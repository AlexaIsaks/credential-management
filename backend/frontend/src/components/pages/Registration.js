import React from "react";
import { withRouter } from "react-router-dom";

// Registration page
class Registration extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: null,
      email: null,
      username: null,
      password: "",
      confirmedPassword: "",
      formErrors: {
        username: null,
        password: null
      },
      error: null
    };
  }

  // Updates state with registration input change
  handleRegistrationInputChange = (event) => {
    const name = event.target.name;
    const value = event.target.value;

    this.setState({
      [name]: value
    });
  }

  // Submits user's registration
  handleRegistrationSubmission = (event) => {
    event.preventDefault();
    const { name, email, username, password, confirmedPassword } = this.state;

    // Password and confirmed password should match
    if (password === confirmedPassword) {
      const user = {
        name,
        email,
        username,
        password
      };

      // Fetch arguments
      const parameters = {
        method: "POST",
        headers: {
          "Content-type": "application/json"
        },
        body: JSON.stringify(user)
      };

      fetch("/registration", parameters)
        .then((response) => response.json())
        .then(
          (results) => {
            if (results.error) {
              this.setState({
                password: "",
                confirmedPassword: "",
                formErrors: results.error
              });
              // Registration successful
            } else {
              this.setState({
                name: null,
                email: null,
                username: "",
                password: "",
                confirmedPassword: ""
              });

              this.props.setLogin(true);

              // Redirect user to portal page
              this.props.history.push({
                pathname: "/portal",
                state: results.token
              });
            }
          },
          (error) => {
            this.setState({
              password: "",
              confirmedPassword: "",
              error
            });
          }
        );
    } else {
      this.setState({
        password: "",
        confirmedPassword: "",
        formErrors: {
          password: "Password and confirmed password do not match."
        }
      });
    }
  };

  render() {
    let displayRegistrationContent = null;

    // An error occurred
    if (this.state.error) {
      displayRegistrationContent = (
        <p className="pt-5 text-center">
          <span className="d-flex justify-content-center align-items-center fs-3 fw-bold text-danger">
            <ion-icon name="sad-outline"></ion-icon>
          </span>
          <span>
            An error has occurred. Please refresh the page or speak to you
            administrator.
          </span>
        </p>
      );
    } else {
      displayRegistrationContent = (
        <div className="row py-5">
          <div className="col-10 col-md-6 col-lg-4 offset-1 offset-md-3 offset-lg-4 py-4 px-3 border">
            <h1 className="h5 text-center">Credential Management</h1>
            <h2 className="mb-5 h5 text-center">Registration</h2>

            {/*Registration form */}
            <form onSubmit={this.handleRegistrationSubmission} className="form">
              <div className="mb-3">
                <label htmlFor="registration-name">Fullname</label>
                <input
                  type="text"
                  name="name"
                  id="registration-name"
                  onChange={this.handleRegistrationInputChange}
                  className="form-control"
                  required
                />
              </div>

              <div className="mb-3">
                <label htmlFor="registration-email">Email</label>
                <input
                  type="text"
                  name="email"
                  id="registration-email"
                  onChange={this.handleRegistrationInputChange}
                  className="form-control"
                  required
                />
              </div>

              <div>
                <label htmlFor="registration-username">Username</label>
                <input
                  type="text"
                  name="username"
                  id="registration-username"
                  onChange={this.handleRegistrationInputChange}
                  className="form-control"
                  required
                />
              </div>

              {/*Username error */}
              {this.state.formErrors.username && (
                <div className="text-danger form__error">
                  {this.state.formErrors.username}
                </div>
              )}

              <div className="my-3">
                <label htmlFor="registration-password">Password</label>
                <input
                  type="password"
                  name="password"
                  value={this.state.password}
                  id="registration-password"
                  autoComplete="off"
                  onChange={this.handleRegistrationInputChange}
                  className="form-control"
                  required
                />
              </div>

              <div>
                <label htmlFor="registration-confirmed-password">
                  Confirm password
                </label>
                <input
                  type="password"
                  name="confirmedPassword"
                  value={this.state.confirmedPassword}
                  id="registration-confirmed-password"
                  autoComplete="off"
                  onChange={this.handleRegistrationInputChange}
                  className="form-control"
                  required
                />
              </div>

              {/*Password error */}
              {this.state.formErrors.password && (
                <div className="text-danger form__error">
                  {this.state.formErrors.password}
                </div>
              )}

              <button
                type="submit"
                className="mt-3 btn btn-outline-secondary w-100"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
      );
    }

    return (
      <main className="pt-4 pt-md-5 min-vh-100 container">
        {/*Main content */}
        {displayRegistrationContent}
      </main>
    );
  }
}

export default withRouter(Registration);
