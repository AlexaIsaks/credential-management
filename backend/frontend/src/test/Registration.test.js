import React from "react";
import Registration from "../components/pages/Registration";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

//Snapshot test: Registration
test("Registration component renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Registration />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
