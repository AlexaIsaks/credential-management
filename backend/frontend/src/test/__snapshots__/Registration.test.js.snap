// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration component renders correctly 1`] = `
<main
  className="pt-4 pt-md-5 min-vh-100 container"
>
  <div
    className="row py-5"
  >
    <div
      className="col-10 col-md-6 col-lg-4 offset-1 offset-md-3 offset-lg-4 py-4 px-3 border"
    >
      <h1
        className="h5 text-center"
      >
        Credential Management
      </h1>
      <h2
        className="mb-5 h5 text-center"
      >
        Registration
      </h2>
      <form
        className="form"
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            htmlFor="registration-name"
          >
            Fullname
          </label>
          <input
            className="form-control"
            id="registration-name"
            name="name"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            htmlFor="registration-email"
          >
            Email
          </label>
          <input
            className="form-control"
            id="registration-email"
            name="email"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div>
          <label
            htmlFor="registration-username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="registration-username"
            name="username"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div
          className="my-3"
        >
          <label
            htmlFor="registration-password"
          >
            Password
          </label>
          <input
            autoComplete="off"
            className="form-control"
            id="registration-password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </div>
        <div>
          <label
            htmlFor="registration-confirmed-password"
          >
            Confirm password
          </label>
          <input
            autoComplete="off"
            className="form-control"
            id="registration-confirmed-password"
            name="confirmedPassword"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </div>
        <button
          className="mt-3 btn btn-outline-secondary w-100"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</main>
`;
